src/geant4/DagSolid.cc

```diff
@@ -0,0 +1,549 @@
+// *                                                                  *
+// * This  code  implementation is the result of  the  scientific and *
+// * technical work of the GEANT4 collaboration and of QinetiQ Ltd,   *
+// * subject to DEFCON 705 IPR conditions.                            *
+// * By using,  copying,  modifying or  distributing the software (or *
+// * any work based  on the software)  you  agree  to acknowledge its *
+// * use  in  resulting  scientific  publications,  and indicate your *
+// * acceptance of all terms of the Geant4 Software license.          *
+// ********************************************************************
+//
+// $Id: DagSolid.hh,v 1.10 2010/12/10 16:30:13 gunter Exp $
+// GEANT4 tag $Name: geant4-09-05 $
+//
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+//
+// MODULE:              DagSolid.hh
+//
+// Date:                27/03/2014
+// Author:              A. Davis M. C. Han, C. H. Kim, J. H. Jeong, Y. S. Yeom, S.
+//                      Kim, Paul. P. H. Wilson, J. Apostolakis
+// Organisation:        The University of Wisconsin-Madison, USA & Hanyang Univ., KR
+//
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+//
+// CHANGE HISTORY
+// --------------
+//
+// 27 March 2014, A Davis, UW - Updated description text
+// 31 October 2010, J. H. Jeong, Hanyang Univ., KR
+//  - Created.
+//
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+//#include "G4TessellatedSolid.hh"
+#include "DagSolid.hh"
+#include "globals.hh"
+#include "Randomize.hh"
+#include "G4TriangularFacet.hh"
+#include "G4VFacet.hh"
+#include "G4TessellatedSolid.hh"
+#include "G4SystemOfUnits.hh"
+
+#include <iostream>
+
+#include "moab/Interface.hpp"
+#include "moab/Range.hpp"
+#include "moab/CartVect.hpp"
+#include "DagMC.hpp"
+using namespace moab;
+
+#include "DagSolid.hh"
+
+#define plot true
+
+//#define G4SPECSDEBUG 1
+///////////////////////////////////////////////////////////////////////////////
+//
+// Standard contructor has blank name and defines no facets.
+//
+DagSolid::DagSolid()
+  : G4TessellatedSolid("dummy"), cubicVolume(0.), surfaceArea(0.) {
+
+  geometryType = "DagSolid";
+
+  xMinExtent =  kInfinity;
+  xMaxExtent = -kInfinity;
+  yMinExtent =  kInfinity;
+  yMaxExtent = -kInfinity;
+  zMinExtent =  kInfinity;
+  zMaxExtent = -kInfinity;
+//G4TessellatedSolid
+//  SetRandomVectors();
+
+  // SetRndVectors();
+
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Alternative constructor. Simple define name and geometry type - no facets
+// to detine.
+//
+DagSolid::DagSolid(const G4String& name, DagMC* dagmc, int volID)
+  : G4TessellatedSolid(name), cubicVolume(0.), surfaceArea(0.) {
+  geometryType = "DagSolid";
+  Myname = name;
+
+  fdagmc = dagmc;
+  fvolID = volID;
+  fvolEntity = fdagmc->entity_by_index(3, volID);
```

Comment:
Optional: Typically you want the constructor of a class to have a very low surface area for errors, to avoid having high compute load or exceptions happen during initialization.  Constructors are called by default when doing common operations like copying vectors of values, so getting this right can have a distinct impact on performance.

A elegant, well-encapsulated constructor typically consists of just declarations and straight value assignments.

Consequently, actions like lookups into pointer objects, for-loops to copy data, etc, are best initiated at the call site or in a subsequent `init()` method that can be called after the constructor is initiated.

I've also seen builder patterns used to great effect [1], or having static member functions that can be called with a constructed object to do the compute-heavy work post-initialization.

In this specific case, I would leverage the default constructor and turn this into a static intializer, rather than having it be an alternative constructor.

[1] https://sourcemaking.com/design_patterns/builder/cpp/1